### PR TITLE
Aliasing based on cycles in CLASSIFY statements

### PR DESCRIPTION
**Pull request by hanjoosten**
This pull request enables aliasing. The idea is to find all concepts that are in a cycle of CLASSIFY paths. These concepts (in the p-structure) will be converted into a single concept in the A-structure. Further processing is as normal. 
A simple testcase is provided as weill.

### CODE REVIEW

**Comment by hanjoosten** [comment]
This will fix https://github.com/AmpersandTarski/Ampersand/issues/899

**Comment by Michiel-s** [comment]
I think it is also important to take a look at the generated json files for the prototype backend. How are they affected? The following attributes are provided for every concept:
- specializations
- generalizations
- directSpecs
- directGens
- largestConcept
How are these different?

Or.. are concepts in classifcation cycles treated as synonyms and will there only be one concept definition in the generated output? In that case I would like to see a new attribute specifying the synonyms..

**Comment by hanjoosten** [comment]
The code in the json files should not have any knowledge about the existence of aliases at all. In the a-structure all aliased concepts are reduced to a single concept. Only in the internal (Haskell) representation of concepts at the A-structure side, the concepts are attributed with a list of aliases. 
The attributes @Michiel-s mentions in the above comment all take into account this reduction of the number of concepts. I expect this all to work as normal. 

Having said that, I mentioned in an earlier comment to this issue that the `show` function on a A_Concept now also shows that aliases, if there are any. If that function is used to produce any content in the json file, that could possibly show the aliases as well. That should not be the case. I'll check this right away, and if so, I will fix it immidiately.

**Comment by hanjoosten** [comment]
I have done some inspection. I will return the `show` function of A_Concept to its original behaviour, so it will only the name of the concept, even if there are aliases. The show function is used at quite a lot of other places, so there might be impact at unexpected places. 
I will creat a new function to show aliases if they exist. This is useful at places where concepts are shown in error messages.

File: src/Ampersand/ADL1/Disambiguate.hs [hunk 1]
```diff
@@ -39,39 +39,42 @@ class Traversable d => Disambiguatable d where
...
---         disambInfo (PCps o a b) (ia1,ib1)
+--         disambInfo cptMap (PCps o a b) (ia1,ib1)
 --            = ( PCps o a' b', (ia, ib) ) -- here only bottom-up information is allowed: don't use ia1 or ib1 here!
---          where (a', (ia,ic1)) = disambInfo a (ia1,ic2) -- here ic2 is top-down, so that is ok
---                (b', (ic2,ib)) = disambInfo b (ic1,ib1)
-  disambInfo :: d (TermPrim,DisambPrim)  --the thing that is disabmiguated
+--          where (a', (ia,ic1)) = disambInfo cptMap a (ia1,ic2) -- here ic2 is top-down, so that is ok
+--                (b', (ic2,ib)) = disambInfo cptMap b (ic1,ib1)
+  disambInfo :: 
+      ConceptMap -- required to turn P_Concepts into proper A_Concepts
```
Comment (stefjoosten) [inline src/Ampersand/ADL1/Disambiguate.hs]: see issue #999

File: src/Ampersand/Input/ADL1/CtxError.hs [hunk 2]
```diff
@@ -174,9 +160,9 @@ mkMultipleRootsError roots gs
              ] ++
              [ "  - "++showA x++" at "++showFullOrig (origin x) | x<-NE.toList gs]++
              [ "Parhaps you could add the following statements:"]++
-             [ "  CLASSIFY "++name cpt++" ISA "++show rootName    | cpt<-roots]
+             [ "  CLASSIFY "++showWithAliases cpt++" ISA "++show rootName    | cpt<-roots]
```
Comment (stefjoosten) [inline src/Ampersand/Input/ADL1/CtxError.hs]: Are you sure that `show rootName` is not one of `showWithAliases cpt`?

Comment (hanjoosten) [reply src/Ampersand/Input/ADL1/CtxError.hs]: Yes, it is part of the definition of the function `rootName`:
~~~
rootName = L.intercalate "_Or_" . map showWithAliases $ roots
~~~

**Pull request merged**